modules/adagioAnalyticsAdapter.js

```diff
@@ -24,24 +42,304 @@ function canAccessTopWindow() {
+    if (deepAccess(adagioAdUnitBids, 'length', 0) <= 0) {
+      logInfo(`Adagio is not on the bid requests for ad unit '${adUnitCode}' and auction '${prebidAuctionId}'`)
+      return;
+    }
+    // Get Adagio params from the first bid.
+    // We assume that all Adagio bids for a same adunit have the same params.
+    const params = adagioAdUnitBids[0].params;
+
+    // Get all media types requested for Adagio.
+    const adagioMediaTypes = removeDuplicates(
+      adagioAdUnitBids.map(bid => Object.keys(bid.mediaTypes)).flat(),
+      mediaTypeKey => mediaTypeKey
+    ).flat().map(mediaType => getMediaTypeAlias(mediaType)).sort();
+
+    const qp = {
+      v: 0,
+      pbjsv: PREBID_VERSION,
+      org_id: params.organizationId,
+      site: params.site,
+      pv_id: params.pageviewId,
+      auct_id: params.adagioAuctionId,
+      adu_code: adUnitCode,
+      url_dmn: w.location.hostname,
+      dvc: params.environment,
+      pgtype: params.pagetype,
+      plcmt: params.placement,
+      mts: mediaTypesKeys.join(','),
+      ban_szs: bannerSizes.join(','),
+      bdrs: bidders.map(bidder => getAdapterNameForAlias(bidder.bidder)).sort().join(','),
+      adg_mts: adagioMediaTypes.join(',')
+    };
+
+    cache.auctions[prebidAuctionId][adUnitCode] = qp;
+    sendNewBeacon(prebidAuctionId, adUnitCode);
+  });
+};
+
+/**
+ * handlerBidResponse allow to track the adagio bid response
+ * and to update the auction cache with the seat ID.
+ * No beacon is sent here.
+*/
+function handlerBidResponse(event) {
+  if (!guard.adagio(event.bidder)) {
+    return;
+  }
+
+  if (!guard.bidTracked(event.auctionId, event.adUnitCode)) {
+    return;
+  }
+
+  cache.updateAuction(event.auctionId, event.adUnitCode, {
+    adg_sid: event.seatId || null
+  });
+};
+
+function handlerBidWon(event) {
+  if (!guard.bidTracked(event.auctionId, event.adUnitCode)) {
+    return;
+  }
+
+  let conversionRate = null;
+  let cmpAdjustmentRate = null;
+  try {
+    if (typeof getGlobal().convertCurrency === 'function') {
+      // Currency module is loaded, we can calculate the conversion rate.
+      // Using new const variables instead of reusing the ones above to avoid confusion.
+
+      // Get the conversion rate from the original currency to USD.
+      const curRateOriginalToUSD = getGlobal().convertCurrency(1, event.originalCurrency, 'USD');
+      conversionRate = curRateOriginalToUSD;
+
+      // Get the cpm from the original currency to the ad server currency, we do that because
+      // event.cpm is the cpm with the adjustment applied.
+      const cpmInAdServerCurrencyBeforeAdjs = getGlobal().convertCurrency(event.originalCpm, event.originalCurrency, event.currency);
+      const adjstRate = event.cpm / cpmInAdServerCurrencyBeforeAdjs;
+      cmpAdjustmentRate = adjstRate;
+    } else {
+      // Currency module is not loaded, we won't be able to calculate the conversion rate.
+      const adjstRate = event.cpm / event.originalCpm;
```

Comment:
So we try to deduce the CPM Adjustment rate ourselves.
But can't Prebid give it to us ?

We cannot be confident in the beacon's data if deductions were made on the client side.

Reply:
Is your question specifically for the `else` case or even for the other one ?
The problem with the adjustment is that it is not necessarily a rate.
From my understanding, it can help publishers to adjust a cpm when the bidder is gross and the publisher wants the net for the auction.

2.2 and 2.3 of the following:
https://docs.prebid.org/dev-docs/publisher-api-reference/bidderSettings.html